crates/red_knot_test/src/matcher.rs

```diff
@@ -0,0 +1,796 @@
+            }
+        }
+        for diagnostic in unmatched {
+            failures.push(diagnostic.unmatched());
+        }
+        if failures.is_empty() {
+            Ok(())
+        } else {
+            Err(failures)
+        }
+    }
+
+    fn column<T: Ranged>(&self, ranged: &T) -> OneIndexed {
+        self.line_index
+            .source_location(ranged.range().start(), &self.source)
+            .column
+    }
+
+    /// Check if `assertion` matches any [`Diagnostic`]s in `unmatched`.
+    ///
+    /// If so, return `true` and remove the matched diagnostics from `unmatched`. Otherwise, return
+    /// `false`.
+    ///
+    /// An `Error` assertion can only match one diagnostic; even if it could match more than one,
+    /// we short-circuit after the first match.
+    ///
+    /// A `Revealed` assertion must match a revealed-type diagnostic, and may also match an
+    /// undefined-reveal diagnostic, if present.
+    fn matches<T: Diagnostic>(&self, assertion: &Assertion, unmatched: &mut Vec<&T>) -> bool {
+        match assertion {
+            Assertion::Error(error) => {
+                let mut found = false;
+                unmatched.retain(|diagnostic| {
+                    if found {
+                        return true;
+                    }
+                    if !error.rule.is_some_and(|rule| rule != diagnostic.rule())
+                        && !error
+                            .column
+                            .is_some_and(|col| col != self.column(*diagnostic))
+                        && !error
+                            .message_contains
+                            .is_some_and(|needle| !diagnostic.message().contains(needle))
+                    {
+                        found = true;
+                        false
+                    } else {
+                        true
+                    }
+                });
+                found
+            }
+            Assertion::Revealed(expected_type) => {
+                let mut matched_revealed_type = None;
+                let mut matched_undefined_reveal = None;
+                for (index, diagnostic) in unmatched.iter().enumerate() {
+                    if matched_revealed_type.is_none()
+                        && diagnostic.rule() == "revealed-type"
+                        && diagnostic.message() == format!("Revealed type is `{expected_type}`")
+                    {
+                        matched_revealed_type = Some(index);
+                        if matched_undefined_reveal.is_some() {
+                            break;
+                        }
+                    } else if matched_undefined_reveal.is_none()
+                        && diagnostic.rule() == "undefined-reveal"
+                    {
+                        matched_undefined_reveal = Some(index);
+                        if matched_revealed_type.is_some() {
+                            break;
+                        }
+                    }
```

Comment:
Nit, nit:
```suggestion
                let mut matched_revealed_type = None;
                let mut matched_undefined_reveal = None;
                let expected_reveal_type_message = format!("Revealed type is `{expected_type}`");
                for (index, diagnostic) in unmatched.iter().enumerate() {
                    if diagnostic.rule() == "revealed-type"
                        && diagnostic.message() == &expected_reveal_type_message
                    {
                        matched_revealed_type = matched_revealed_type.or(Some(index));
                    } else if diagnostic.rule() == "undefined-reveal" {
                        matched_undefined_reveal = matched_undefined_reveal.or(Some(index));
                    }

                    if matched_revealed_type.is_some() && matched_undefined_reveal.is_some() {
                        break;
                    }
                }
```

Reply:
Ah, thanks for catching the `format!` in loop, that was dumb.

The one bit I didn't take here is removing the `is_none()` checks; I see that `.or(...)` also works fine, and I suspect the perf doesn't matter either way, but I find it a bit clearer to read the logic if we are explicit that we don't need to look for the diagnostics that we already found. The `.or()` is a bit more subtle that we look again, but just don't replace the index if we already have one.